tests/conformance/test_conformance.py

```diff
@@ -0,0 +1,699 @@
+    See also: https://github.com/googleapis/google-cloud-cpp/tree/main/google/cloud/storage/emulator
+    """
+    import json
+
+    preflight_post_uri = host + "/retry_test"
+    headers = {
+        "Content-Type": "application/json",
+    }
+    data_dict = {"instructions": {method_name: instructions}}
+    data = json.dumps(data_dict)
+    r = requests.post(preflight_post_uri, headers=headers, data=data)
+    return r.json()
+
+
+def _get_retry_test(host, id):
+    status_get_uri = "{base}{retry}/{id}".format(base=host, retry="/retry_test", id=id)
+    r = requests.get(status_get_uri)
+    return r.json()
+
+
+def _run_retry_test(
+    host, id, lib_func, _preconditions, bucket, object, notification, hmac_key
+):
+    """
+    To execute tests against the list of instrucions sent to the Retry API, create a client to send the retry test ID using the x-retry-test-id header in each request.
+    For incoming requests which match the given API method, the emulator will pop off the next instruction from the list and force the listed failure case.
+    """
+    client = storage.Client(client_options={"api_endpoint": host})
+    client._http.headers.update({"x-retry-test-id": id})
+    lib_func(
+        client,
+        _preconditions,
+        bucket=bucket,
+        object=object,
+        notification=notification,
+        hmac_key=hmac_key,
+    )
+
+
+def _delete_retry_test(host, id):
+    status_get_uri = "{base}{retry}/{id}".format(base=host, retry="/retry_test", id=id)
+    requests.delete(status_get_uri)
+
+
+########################################################################################################################################
+### Run Test Case for Retry Strategy ###################################################################################################
+########################################################################################################################################
+
+
+def run_test_case(
+    scenario_id, method, case, lib_func, host, bucket, object, notification, hmac_key
+):
+    scenario = _CONFORMANCE_TESTS[scenario_id - 1]
+    expect_success = scenario["expectSuccess"]
+    precondition_provided = scenario["preconditionProvided"]
+    method_name = method["name"]
+    instructions = case["instructions"]
+
+    try:
+        r = _create_retry_test(host, method_name, instructions)
+        id = r["id"]
+    except Exception as e:
+        raise Exception(
+            "Error creating retry test for {}: {}".format(method_name, e)
+        ).with_traceback(e.__traceback__)
+
+    # Run retry tests on library methods.
+    try:
+        _run_retry_test(
+            host,
+            id,
+            lib_func,
+            precondition_provided,
+            bucket,
+            object,
+            notification,
+            hmac_key,
+        )
+    except Exception as e:
```

Comment:
Why catch and log this instead of just letting the exception bubble up and have pytest deal with it?

Reply:
Catching the exception here helps verify whether or not the test ran successfully, with a variable `success_results`. `success_results` is compared against `expect_success` from the schema. 

For test scenarios 3 and 4, the tests are expected to fail, hence why we're catching the exception.

Reply:
I see-- would the test logs be filled with all these exceptions?

Reply:
Yep the exceptions here are caught and logged.